L1Trigger/Phase2L1ParticleFlow/plugins/L1HPSPFTauProducer.cxx

```diff
@@ -0,0 +1,204 @@
+
+  static std::vector<L1HPSPFTauEmu::Particle> convertJetsToHW(std::vector<edm::Ptr<reco::CaloJet>>& edmJets); 
+  static std::vector<L1HPSPFTauEmu::Particle> convertEDMToHW(std::vector<edm::Ptr<l1t::PFCandidate>>& edmParticles);
+  static std::vector<l1t::PFTau> convertHWToEDM(std::vector<L1HPSPFTauEmu::Tau> hwTaus);
+  
+};
+
+L1HPSPFTauProducer::L1HPSPFTauProducer(const edm::ParameterSet& cfg)
+   : _nTaus(cfg.getParameter<int>("nTaus")),
+     _HW(cfg.getParameter<bool>("HW")),
+     fUseJets_(cfg.getParameter<bool>("useJets")),
+     _debug(cfg.getParameter<bool>("debug")){ //,
+      srcL1PFCands_ = cfg.getParameter<edm::InputTag>("srcL1PFCands");
+      tokenL1PFCands_ = consumes<l1t::PFCandidateCollection>(srcL1PFCands_);   
+      
+      srcL1PFJets_ = cfg.getParameter<edm::InputTag>("srcL1PFJets");
+      tokenL1PFJets_ = consumes<std::vector<reco::CaloJet>>(srcL1PFJets_); 
+      produces<l1t::PFTauCollection>("HPSTaus");
+      
+      produces<l1t::PFCandidateCollection>("SelPFCands");
+
+     }
+
+void L1HPSPFTauProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions){
+  edm::ParameterSetDescription desc;    
+  
+  desc.add<edm::InputTag>("srcL1PFCands", edm::InputTag("l1tLayer1","Puppi"));    
+  desc.add<int>("nTaus", 16);
+  desc.add<bool>("HW", true);
+  desc.add<bool>("useJets",false);
+  desc.add<bool>("debug", false);   
+  desc.add<edm::InputTag>("srcL1PFJets", edm::InputTag("l1tPhase1JetProducer","UncalibratedPhase1L1TJetFromPfCandidates"));
+  descriptions.add("L1HPSPFTauProducer", desc);
+}
+
+
+void L1HPSPFTauProducer::produce(edm::StreamID,
+                                 edm::Event& iEvent,
+                                 const edm::EventSetup& iSetup) const {
+
+  std::unique_ptr<l1t::PFTauCollection> newPFTauCollection(new l1t::PFTauCollection);
+  std::unique_ptr<l1t::PFCandidateCollection> selParticles(new l1t::PFCandidateCollection);
+  edm::Handle<l1t::PFCandidateCollection> l1PFCandidates;
+  iEvent.getByToken(tokenL1PFCands_, l1PFCandidates);
+  //add jets even if not used, for simplicity
+  edm::Handle<std::vector<reco::CaloJet>> l1PFJets;
+  iEvent.getByToken(tokenL1PFJets_, l1PFJets);
+  //
+
+  //adding collection
+  std::vector<edm::Ptr<l1t::PFCandidate>> particles;
+  for(unsigned i = 0; i < (*l1PFCandidates).size(); i++) {
+
+    particles.push_back(edm::Ptr<l1t::PFCandidate>(l1PFCandidates, i));  
+    
+  }
+
+
+  //get the jets
+  std::vector<edm::Ptr<reco::CaloJet>> jets;
+  for(unsigned int i = 0; i < (*l1PFJets).size(); i++){
+    jets.push_back(edm::Ptr<reco::CaloJet>(l1PFJets, i));  
+  //
+  }
+
+  std::vector<l1t::PFTau> taus;
+ 
+  taus = processEvent_HW(particles, jets);
+  
+
+  std::sort(taus.begin(), taus.end(), [](l1t::PFTau i, l1t::PFTau j) { return (i.pt() > j.pt()); });
+  newPFTauCollection->swap(taus);
+  iEvent.put(std::move(newPFTauCollection), "HPSTaus");
```

Comment:
If you would use the `EDPutTokenT`, the `newPFTauCollection` could be avoided along
```suggestion
  std::vector<l1t::PFTau> taus = processEvent_HW(particles, jets);
  std::sort(taus.begin(), taus.end(), [](l1t::PFTau i, l1t::PFTau j) { return (i.pt() > j.pt()); });
  iEvent.emplace(pfTauPutToken_, std::move(taus));
```

Reply:
May I also suggest to pass const reference to object instead of the object itself, wherever applicable, e.g.

std::sort(taus.begin(), taus.end(), [](l1t::PFTau i, l1t::PFTau j) { return (i.pt() > j.pt()); });

may be changed to 
std::sort(taus.begin(), taus.end(), [](const l1t::PFTau& i, const l1t::PFTau& j) { return (i.pt() > j.pt()); });